### PR TITLE
Permission checks on controller and methods

### PR DESCRIPTION
**Pull request by qifeng-bai**
Two issues (#191) were found:

1, Permissions added on the method was overwritten by those on the controller
2, When a signed-in user is refused due to no sufficient permissions, it will be redirected to 'auth.ala.org.au'. Since the user is already signed in, Auth.ala.org.au redirects to the previous link. Then it goes another loop.

### CODE REVIEW

File: grails-app/controllers/au/org/ala/spatial/service/LoginInterceptor.groovy [hunk 1]
```diff
@@ -39,21 +39,40 @@ class LoginInterceptor {
...
                 return accessDenied(STATUS_UNAUTHORISED,'Forbidden, ApiKey or user login required!')
             }
-        }
-        // UserId is required
-        // May need to check role
-        if (method?.isAnnotationPresent(RequireAdmin) || controllerClass?.isAnnotationPresent(RequireAdmin)) {
+        } else if ( permissionLevel == RequireAdmin ) {
             if (serviceAuthService.hasValidApiKey())
                 return true
             role = grailsApplication.config.auth.admin_role //recommended: ROLE_ADMIN
```
Comment (nickdos) [inline grails-app/controllers/au/org/ala/spatial/service/LoginInterceptor.groovy:74]: Please change `grailsApplication.config.auth.admin_role` to be `grailsApplication.config.getProperty('auth.admin_role', String, 'ROLE_ADMIN')` so it is type safe and has fallback default value. `grailsApplication.config.getProperty()` should always be used to read config vars...

File: grails-app/controllers/au/org/ala/spatial/service/LoginInterceptor.groovy [hunk 2]
```diff
@@ -39,21 +39,40 @@ class LoginInterceptor {
...
+                permissionLevel = RequireLogin
+            } else if (controllerClass?.isAnnotationPresent(RequireAdmin)){
+                permissionLevel = RequireAdmin
+            }
+        }
+
+        //Permission check
         def role  // if require a certain level of ROLE
-        if ( method?.isAnnotationPresent(RequirePermission) || controllerClass?.isAnnotationPresent(RequirePermission) ) {
+        if ( permissionLevel == RequirePermission ) {
```
Comment (nickdos) [inline grails-app/controllers/au/org/ala/spatial/service/LoginInterceptor.groovy:65]: Fix whitespace in `if` pls

File: grails-app/controllers/au/org/ala/spatial/service/LoginInterceptor.groovy [hunk 3]
```diff
@@ -39,21 +39,40 @@ class LoginInterceptor {
             return true
         }
 
+        //Calculating the required permission.
+        def permissionLevel = null;
+        //Permission on method has the top priority
+        if ( method?.isAnnotationPresent(RequirePermission)) {
```
Comment (nickdos) [inline grails-app/controllers/au/org/ala/spatial/service/LoginInterceptor.groovy:45]: whitespace is inconsistent - remove `( method` space

**Review by nickdos — CHANGES_REQUESTED** [review]
few changes

File: grails-app/controllers/au/org/ala/spatial/service/LoginInterceptor.groovy (hunk 2 above)
Comment (qifeng-bai) [reply grails-app/controllers/au/org/ala/spatial/service/LoginInterceptor.groovy:65]: Done

File: grails-app/controllers/au/org/ala/spatial/service/LoginInterceptor.groovy (hunk 3 above)
Comment (qifeng-bai) [reply grails-app/controllers/au/org/ala/spatial/service/LoginInterceptor.groovy:45]: Done

File: grails-app/controllers/au/org/ala/spatial/service/LoginInterceptor.groovy (hunk 1 above)
Comment (qifeng-bai) [reply grails-app/controllers/au/org/ala/spatial/service/LoginInterceptor.groovy:74]: Done

File: grails-app/controllers/au/org/ala/spatial/service/LoginInterceptor.groovy (hunk 3 above)
Comment (qifeng-bai) [reply grails-app/controllers/au/org/ala/spatial/service/LoginInterceptor.groovy:45]: reformatted

**Comment by qifeng-bai** [comment]
Tests:

Go to: spatial-test.ala.org.au -> Click "Copy layers from remote servers" 
1, if you are a ROLE_USER, you should see: "You do not have permission to access this link. Please sign in with another account."

2, if you are not logged in,  you should see:  "sign in to access this link", and page will be redirect to login page after 3 seconds

You should see the same responses when you access : "http://devt.ala.org.au:8080/ws/manageLayers/layers" -> click 'Edit'

File: grails-app/controllers/au/org/ala/spatial/service/LoginInterceptor.groovy (hunk 3 above)
Comment (qifeng-bai) [reply grails-app/controllers/au/org/ala/spatial/service/LoginInterceptor.groovy]: reformatted

**Review by nickdos — APPROVED** [review]

**Pull request merged**